python/examples/extensions/Custom_Metrics.ipynb

```diff
@@ -0,0 +1,216 @@
+        "                    self.histogram.value.update(arr)\n",
+        "\n",
+        "        for lst in [data.list.ints, data.list.floats]:\n",
+        "            if lst is not None and len(lst) > 0:\n",
+        "                self.histogram.value.update_list(lst)\n",
+        "\n",
+        "        return OperationResult.ok(successes)\n",
+        "\n",
+        "    # The zero method returns an \"empty\" instance of your metric ready to start tracking data\n",
+        "    # If your metric needs configuration, create a subclass of MetricConfig containing your\n",
+        "    # parameters.\n",
+        "    @classmethod\n",
+        "    def zero(cls, config: MetricConfig) -> \"HistogramMetric\":\n",
+        "        return cls(histogram=KllComponent(ds.kll_doubles_sketch(k=config.kll_k)))\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "DLXIn7m471TF"
+      },
+      "source": [
+        "If you prefer not to use `MetricComponent` fields for your metric, you can instead make your metric a subclass of `CustomMetricBase`. All fields whose names don't start with `_` will be included in the metric summary and serialized via protobuf."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "8ie0jJoI9FHv"
+      },
+      "outputs": [],
+      "source": [
+        "@dataclass\n",
+        "class StructMetric(CustomMetricBase):\n",
+        "    x: int\n",
+        "    s: str\n",
+        "    _private: float = 3.14159  # excluded from summary and protobuf\n",
+        "\n",
+        "    @property\n",
+        "    def namespace(self) -> str:\n",
+        "        return \"struct\"\n",
+        "\n",
+        "    # you must implement your own merge method\n",
+        "    def merge(self, other: \"StructMetric\") -> \"StructMetric\":\n",
+        "        return StructMetric(self.x + other.x, self.s + other.s)\n",
+        "\n",
+        "    def columnar_update(self, data: PreprocessedColumn) -> OperationResult:\n",
+        "        self.x += 1\n",
+        "        self.s += \"a\"\n",
+        "        return OperationResult.ok(1)\n",
+        "\n",
+        "    @classmethod\n",
+        "    def zero(cls, config: MetricConfig) -> \"StructMetric\":\n",
+        "        return cls(0, \"\")\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "-zZnKNCF9oUj"
+      },
+      "source": [
+        "# Using Your Metric\n",
+        "You will need to create a `Resolver` and `DatasetSchema` in order to use your metric."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "HlFNH-H6-qKg"
+      },
+      "outputs": [],
+      "source": [
+        "class TestResolver(Resolver):\n",
+        "    def resolve(self, name: str, why_type: DataType, column_schema: ColumnSchema) -> Dict[str, Metric]:\n",
```

Comment:
This line giving me an error when run in colab, are you missing:
```from whylogs.core import ColumnSchema```

```NameError                                 Traceback (most recent call last)
[<ipython-input-4-7cba2972d2b3>](https://localhost:8080/#) in <module>()
----> 1 class TestResolver(Resolver):
      2     def resolve(self, name: str, why_type: DataType, column_schema: ColumnSchema) -> Dict[str, Metric]:
      3         return {"histogram": HistogramMetric.zero(column_schema.cfg),
      4                 "struct": StructMetric(0, "")}
      5 

[<ipython-input-4-7cba2972d2b3>](https://localhost:8080/#) in TestResolver()
      1 class TestResolver(Resolver):
----> 2     def resolve(self, name: str, why_type: DataType, column_schema: ColumnSchema) -> Dict[str, Metric]:
      3         return {"histogram": HistogramMetric.zero(column_schema.cfg),
      4                 "struct": StructMetric(0, "")}
      5 

NameError: name 'ColumnSchema' is not defined```

Reply:
this was fixed in the last commit